File: BackEnd-Siukat/services/kendaraan.go
package services

import (
	"BackEnd-Siukat/config"
	"BackEnd-Siukat/models"
	"time"
)

type KendaraanService struct{}

func (s *KendaraanService) Add(req models.Kendaraan, atribut string) (models.Kendaraan, error) {
	req.Atribut = atribut
	err := config.DB.Create(&req).Error
	return req, err
}

func (s *KendaraanService) Edit(req models.Kendaraan, noPeserta string, existing models.Kendaraan, atribut string) (models.Kendaraan, error) {
	data := map[string]interface{}{}

	data["status_motor"] = req.StatusMotor
	if req.StatusMotor == "tidak" {
		data["jumlah_motor"] = 0
		data["pajak_motor"] = 0
		data["scan_motor"] = ""
	} else {
		data["jumlah_motor"] = req.JumlahMotor
		data["pajak_motor"] = req.PajakMotor
		if req.ScanMotor != "" {
			data["scan_motor"] = req.ScanMotor
		} else {
			data["scan_motor"] = existing.ScanMotor
		}
	}

	data["status_mobil"] = req.StatusMobil
	if req.StatusMobil == "tidak" {
		data["jumlah_mobil"] = 0
		data["pajak_mobil"] = 0
		data["scan_mobil"] = ""
	} else {
		data["jumlah_mobil"] = req.JumlahMobil
		data["pajak_mobil"] = req.PajakMobil
		if req.ScanMobil != "" {
			data["scan_mobil"] = req.ScanMobil
		} else {
			data["scan_mobil"] = existing.ScanMobil
		}
	}

	if err := config.DB.Model(&models.Kendaraan{}).Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).Updates(data).Error; err != nil {
		return models.Kendaraan{}, err
	}
	var res models.Kendaraan
	config.DB.Where("no_peserta = ? AND atribut = ?", noPeserta, atribut).First(&res)
	return res, nil
}

func (s *KendaraanService) AddLog(user models.Kendaraan, atribut string, executor string, timestamp *time.Time) error {
	logKen := models.LogKendaraan{
		NoPeserta:   user.NoPeserta,
		StatusMotor: user.StatusMotor,
		JumlahMotor: user.JumlahMotor,
		PajakMotor:  user.PajakMotor,
		ScanMotor:   user.ScanMotor,
		StatusMobil: user.StatusMobil,
		JumlahMobil: user.JumlahMobil,
		PajakMobil:  user.PajakMobil,
		ScanMobil:   user.ScanMobil,
		Atribut:     atribut,
		Executor:    executor,
		Timestamp:   timestamp,
	}
	return config.DB.Create(&logKen).Error
}

func (s *KendaraanService) GetByLoggedIn(noPeserta string) (models.Kendaraan, error) {
	var res models.Kendaraan
	err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error
	return res, err
}

func (s *KendaraanService) CheckData(noPeserta string) (bool, error) {
	var res models.Kendaraan
	if err := config.DB.Where("no_peserta = ?", noPeserta).First(&res).Error; err != nil {
		return false, err
	}

	// Replicating Node.js "delete and check" logic using a map
	data := map[string]interface{}{
		"no_peserta":   res.NoPeserta,
		"status_motor": res.StatusMotor,
		"jumlah_motor": res.JumlahMotor,
		"pajak_motor":  res.PajakMotor,
		"scan_motor":   res.ScanMotor,
		"status_mobil": res.StatusMobil,
		"jumlah_mobil": res.JumlahMobil,
		"pajak_mobil":  res.PajakMobil,
		"scan_mobil":   res.ScanMobil,
		"atribut":     res.Atribut,
	}

	if res.StatusMotor == "tidak" {
		delete(data, "jumlah_motor")
		delete(data, "pajak_motor")
		delete(data, "scan_motor")
	}

	if res.StatusMobil == "tidak" {
		delete(data, "jumlah_mobil")
		delete(data, "pajak_mobil")
		delete(data, "scan_mobil")
	}

	for _, val := range data {
		if val == nil || val == "" || val == 0 || val == "0" {
			return false, nil
		}
	}

	return true, nil
}

File: Be-siukat-new/siukat-service-dev_snmptn_2022/services/kendaraan.js
var Kendaraan = require('../models/kendaraan');
var LogKendaraan = require('../models/log_kendaraan');

var kendaraan = function () {};

kendaraan.prototype.add = function(req, atribut){
  return new Promise(function(resolve, reject){
    Kendaraan.create({
      no_peserta: req.no_peserta,
      status_motor: req.status_motor,
      jumlah_motor: req.jumlah_motor,
      pajak_motor: req.pajak_motor,
      scan_motor: req.scan_motor,
      status_mobil: req.status_mobil,
      jumlah_mobil: req.jumlah_mobil,
      pajak_mobil: req.pajak_mobil,
      scan_mobil: req.scan_mobil,
      atribut: atribut
    },{
      fields: [
        'no_peserta',
        'status_motor',
        'jumlah_motor',
        'pajak_motor',
        'scan_motor',
        'status_mobil',
        'jumlah_mobil',
        'pajak_mobil',
        'scan_mobil',
        'atribut'
      ]
    }).then(function(response){
      resolve(response)
    }).catch(function(err){
      reject(err)
    })
  })
}

kendaraan.prototype.edit = function(req, no_peserta, users, atribut){
  return new Promise(function(resolve, reject){
    var data = {
      status_motor: req.body.status_motor,
      jumlah_motor: (req.body.status_motor=="tidak")? "":req.body.jumlah_motor,
      pajak_motor: (req.body.status_motor=="tidak")? "":req.body.pajak_motor,
      scan_motor: users.scan_motor,
      status_mobil: req.body.status_mobil,
      jumlah_mobil: (req.body.status_mobil=="tidak")? "":req.body.jumlah_mobil,
      pajak_mobil: (req.body.status_mobil=="tidak")? "":req.body.pajak_mobil,
      scan_mobil: users.scan_mobil,
    }

    if(req.files["file_scan_motor"] !== undefined){
      data.scan_motor = (req.body.status_motor=="tidak")? "":req.files["file_scan_motor"][0].filename
    }else{
      data.scan_motor = (req.body.status_motor=="tidak")? "":users.scan_motor      
    }
    if(req.files["file_scan_mobil"] !== undefined){
      data.scan_mobil = (req.body.status_mobil=="tidak")? "":req.files["file_scan_mobil"][0].filename
    }else{
      data.scan_mobil = (req.body.status_mobil=="tidak")? "":users.scan_mobil
    }
    Kendaraan.update(data, {
      where: {
        no_peserta: no_peserta, 
		atribut: atribut
      }
    }).then(function(response){
      resolve(response)
    }).catch(function(err){
      reject(err)
    })
  })
}

kendaraan.prototype.addLog = function(users, atribut, executor, timestamp){
  return new Promise(function(resolve, reject){
    LogKendaraan.create({
      no_peserta: users.no_peserta,
      status_motor: users.status_motor,
      jumlah_motor: users.jumlah_motor,
      pajak_motor: users.pajak_motor,
      scan_motor: users.scan_motor,
      status_mobil: users.status_mobil,
      jumlah_mobil: users.jumlah_mobil,
      pajak_mobil: users.pajak_mobil,
      scan_mobil: users.scan_mobil,
      atribut: atribut,
      executor: executor,
      timestamp: timestamp
    }, {
      field: [
        'no_peserta',
        'status_motor',
        'jumlah_motor',
        'pajak_motor',
        'scan_motor',
        'status_mobil',
        'jumlah_mobil',
        'pajak_mobil',
        'scan_mobil',
        'atribut',
        'executor',
        'atribut'
      ]
    }).then(function(response){
      resolve(response)
    }).catch(function(err){
      reject(err)
    })
  })
}

kendaraan.prototype.getByLoggedIn = function(no_peserta){
  return new Promise(function(resolve, reject){
    Kendaraan.findOne({
      where: {
        no_peserta: no_peserta
      }
    }).then(function(response){
      resolve(response)
    }).catch(function(err){
      reject(err)
    })
  })
}

kendaraan.prototype.checkData = function(no_peserta){
  return new Promise((resolve, reject) => {
    Kendaraan.findOne({
      where: {
        no_peserta: no_peserta
      },
      raw: true,
    }).then(function (response){
      var result = true
      if(response.status_motor==="tidak"){
        delete response.jumlah_motor
        delete response.pajak_motor
        delete response.scan_motor
      }

      if(response.status_mobil==="tidak"){
        delete response.jumlah_mobil
        delete response.pajak_mobil
        delete response.scan_mobil
      }

      for(var key in response){
        result *= (response[key] !== 0 && response[key] !== "")
      }
      resolve(result)
    }).catch(function (err) {
      reject(err)
    })
  })
}

module.exports = kendaraan
